contracts/src/oracle/library.cairo

```diff
@@ -226,37 +237,71 @@ namespace Oracle {
         return (threshold,);
     }
 
+    func get_future_entry{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        pair_id, expiry_timestamp, source
+    ) -> (future_entry: FutureEntry) {
+        let (future_entry) = Oracle__future_entry_storage.read(pair_id, expiry_timestamp, source);
+        return (future_entry,);
+    }
+
     //
     // Setters
     //
 
-    func publish_spot_entry{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-        new_entry: SpotEntry
+    func publish_future_entry{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        new_entry: FutureEntry
     ) {
-        alloc_locals;
+        let (entry_ptr: FutureEntry*) = alloc();
+        assert entry_ptr[0] = new_entry;
+        validate_sender_for_source(cast(entry_ptr, felt*));
 
-        let (publisher_registry_address) = get_publisher_registry_address();
-        let (publisher_address) = IPublisherRegistry.get_publisher_address(
-            publisher_registry_address, new_entry.base.publisher
-        );
-        let (_can_publish_source) = IPublisherRegistry.can_publish_source(
-            publisher_registry_address, new_entry.base.publisher, new_entry.base.source
+        let (entry) = Oracle__future_entry_storage.read(
+            new_entry.pair_id, new_entry.expiry_timestamp, new_entry.base.source
         );
 
-        let (caller_address) = get_caller_address();
+        with_attr error_message("Oracle: Existing entry is more recent") {
+            assert_le(entry.base.timestamp, new_entry.base.timestamp);
+        }
 
-        with_attr error_message("Oracle: Publisher and caller must not be 0 addresses") {
-            assert_not_zero(publisher_address);
-            assert_not_zero(caller_address);
+        let (current_timestamp) = get_block_timestamp();
+        with_attr error_message("Oracle: New entry timestamp is too far in the past") {
+            assert_le(current_timestamp - TIMESTAMP_BUFFER, new_entry.base.timestamp);
```

Comment:
do we want to have a lot of places where we `cast(FutureEntry, felt*)` then `cast(felt*, BaseEntry)` ?  There's no generics so this requires this abstraction, which I did for the `validate_sender_for_source`.  But it doesn't help verbosity much to do this since we're abstraction 3 lines of code for at least 6.

Reply:
technically we can individually cast entries to BaseEntry and then handle the generic evaluations, but it might make our code difficult to interpret.  I can add it for this function and see, but in general copying and pasting isn't horrible since we don't have very clean generic implementations, and casting is much less explicit.

Reply:
factored it out, but my thought is that copying and pasting isn't bad when there's a lack of generics.  It was actually recommended by the golang core team in the early days of their language.